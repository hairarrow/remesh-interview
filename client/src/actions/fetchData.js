import "cross-fetch/polyfill";

export const REQUEST_SESSION = "REQUEST_SESSION";
export const requestSession = () => ({ type: REQUEST_SESSION });

export const REQUEST_USERS = "REQUEST_USERS";
export const requestUsers = () => ({ type: REQUEST_USERS });

export const REQUEST_QUESTIONS = "REQUEST_QUESTIONS";
export const requestQuestions = () => ({ type: REQUEST_QUESTIONS });

export const REQUEST_VOTES = "REQUEST_VOTES";
export const requestVotes = () => ({ type: REQUEST_VOTES });

export const REQUEST_MESSAGES = "REQUEST_MESSAGES";
export const requestMessages = () => ({ type: REQUEST_MESSAGES });

export const RECEIVE_SESSION = "RECEIVE_SESSION";
const receiveSession = session => ({
  type: RECEIVE_SESSION,
  session
});

export const RECEIVE_USERS = "RECEIVE_USERS";
const receiveUsers = users => ({
  type: RECEIVE_USERS,
  users
});

export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";
const receiveQuestions = questions => ({
  type: RECEIVE_QUESTIONS,
  questions
});

export const RECEIVE_MESSAGES = "RECEIVE_MESSAGES";
const receiveMessages = messages => ({
  type: RECEIVE_MESSAGES,
  messages
});

export const RECEIVE_VOTES = "RECEIVE_VOTES";
const receiveVotes = votes => ({
  type: RECEIVE_VOTES,
  votes
});

const fetchData = () => dispatch => {
  dispatch(requestMessages());
  dispatch(requestQuestions());
  dispatch(requestVotes());
  dispatch(requestUsers());
  Promise.all([
    fetch("/api/users"),
    fetch("/api/questions"),
    fetch("/api/messages"),
    fetch("/api/votes")
  ])
    .then(
      ([...data]) => Promise.all([...data].map(res => res.json())),
      err => console.log(err)
    )
    .then(([users, questions, messages, votes]) => {
      dispatch(receiveMessages(messages));
      dispatch(receiveVotes(votes));
      dispatch(receiveQuestions(questions));
      dispatch(receiveUsers(users));
      dispatch(
        receiveSession({
          users,
          questions,
          messages,
          votes
        })
      );
    });
};

export const getData = () => dispatch => dispatch(fetchData());
