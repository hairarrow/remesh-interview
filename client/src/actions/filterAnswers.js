export const ADD_FILTER_TO_QUESTION = "ADD_FILTER_TO_QUESTION";
export const addFilterToQuestion = messages => ({
  type: ADD_FILTER_TO_QUESTION,
  messages
});
