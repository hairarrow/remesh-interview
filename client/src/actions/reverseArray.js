export const REVERSE_ARRAY = "REVERSE_ARRAY";

export const reverseArray = () => dispatch => {
  dispatch({
    type: REVERSE_ARRAY,
    payload: "result"
  });
};
