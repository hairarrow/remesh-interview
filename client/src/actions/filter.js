import { filterSession } from "./session";

export const RECEIVE_FILTERS = "RECEIVE_FILTERS";
export const receiveFilters = filters => ({
  type: RECEIVE_FILTERS,
  filters
});

export const toggleFilterOnSession = (FILTER, filters) => dispatch => {
  const { filter, metric, enabled } = FILTER;
  const updateMetric = [...filters[filter]].map(
    fi =>
      fi.metric === metric
        ? {
            ...fi,
            enabled: !enabled
          }
        : { ...fi }
  );
  dispatch(updateFilters({ ...filters, [filter]: updateMetric }));
};

const updateFilters = filters => dispatch => {
  dispatch(receiveFilters(filters));
  dispatch(filterSession());
};

const createFilters = users => dispatch => {
  const { id, ...userModel } = users[0];
  const filters = Object.keys(userModel)
    .map(metrics => ({
      [metrics]: [...new Set(users.map(u => u[metrics]))].map(metric => ({
        metric,
        enabled: false
      }))
    }))
    .reduce((a, b) => ({ ...a, ...b }), {});
  dispatch(receiveFilters(filters));
};

export const getFilters = users => dispatch => {
  dispatch(createFilters(users));
};
