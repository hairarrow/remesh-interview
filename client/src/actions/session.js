import "cross-fetch/polyfill";
import { getFilters } from "./filter";

export const REQUEST_SESSION = "REQUEST_SESSION";
export const requestSession = () => ({ type: REQUEST_SESSION });

export const REQUEST_USERS = "REQUEST_USERS";
export const requestUsers = () => ({ type: REQUEST_USERS });

export const REQUEST_QUESTIONS = "REQUEST_QUESTIONS";
export const requestQuestions = () => ({ type: REQUEST_QUESTIONS });

export const REQUEST_VOTES = "REQUEST_VOTES";
export const requestVotes = () => ({ type: REQUEST_VOTES });

export const REQUEST_MESSAGES = "REQUEST_MESSAGES";
export const requestMessages = () => ({ type: REQUEST_MESSAGES });

export const RECEIVE_SESSION = "RECEIVE_SESSION";
const receiveSession = session => ({
  type: RECEIVE_SESSION,
  session
});

export const RECEIVE_USERS = "RECEIVE_USERS";
const receiveUsers = users => ({
  type: RECEIVE_USERS,
  users
});

export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";
const receiveQuestions = questions => ({
  type: RECEIVE_QUESTIONS,
  questions
});

export const RECEIVE_MESSAGES = "RECEIVE_MESSAGES";
const receiveMessages = messages => ({
  type: RECEIVE_MESSAGES,
  messages
});

export const RECEIVE_VOTES = "RECEIVE_VOTES";
const receiveVotes = votes => ({
  type: RECEIVE_VOTES,
  votes
});

export const FILTER_SESSION = "FILTER_SESSION";
export const filterSession = () => (dispatch, getState) => {
  const { session, filter } = getState();
  const { questions, messages, users, votes } = session;
  const { filters } = filter;

  const enabledFilters = Object.keys(filters)
    .map(metric => ({
      [metric]: [...filters[metric]].filter(({ enabled }) => enabled === true)
    }))
    .reduce((a, b) => {
      if (b[Object.keys(b)[0]].length === 0) return { ...a };
      return { ...a, ...b };
    }, {});

  let filteredUsers = [...users];
  if (Object.keys(enabledFilters).length) {
    filteredUsers = Object.keys(enabledFilters)
      .map(metric =>
        [...users].filter(user => {
          const fis = enabledFilters[metric].map(fi => fi.metric);
          return (
            messages.map(({ creatorId }) => creatorId).includes(user.id) ||
            fis.includes(user[metric])
          );
        })
      )
      .reduce((a, b) => [...a, ...b], []);
  }

  const filteredVotes = [...votes].filter(({ userId }) =>
    [...filteredUsers].map(({ id }) => id).includes(userId)
  );

  dispatch(
    createSession({
      users: filteredUsers,
      votes: filteredVotes,
      questions,
      messages
    })
  );
};

export const getData = () => dispatch => dispatch(fetchData());

export const createSession = ({
  questions,
  messages,
  users,
  votes
}) => dispatch => {
  const organizedSession = [...questions].map(q =>
    Object.assign({}, q, {
      messages: [...messages]
        .filter(({ questionId }) => questionId === q.id)
        .map(({ creatorId, ...msg }) => ({
          ...msg,
          creator: [...users].filter(({ id }) => id === creatorId)[0],
          votes: [...votes]
            .filter(
              ({ questionId, messageId }) =>
                questionId === q.id && messageId === msg.id
            )
            .map(({ id, userId }) =>
              Object.assign(
                {},
                { id },
                [...users].filter(user => user.id === userId)[0]
              )
            )
        }))
    })
  );

  dispatch(receiveSession(organizedSession));
};

const fetchData = () => dispatch => {
  dispatch(requestMessages());
  dispatch(requestQuestions());
  dispatch(requestVotes());
  dispatch(requestUsers());
  Promise.all([
    fetch("/api/users"),
    fetch("/api/questions"),
    fetch("/api/messages"),
    fetch("/api/votes")
  ])
    .then(
      ([...data]) => Promise.all([...data].map(res => res.json())),
      err => console.log(err)
    )
    .then(([users, questions, messages, votes]) => {
      dispatch(receiveMessages(messages));
      dispatch(receiveVotes(votes));
      dispatch(receiveQuestions(questions));
      dispatch(receiveUsers(users));
      dispatch(getFilters(users));
      dispatch(
        createSession({
          users,
          questions,
          messages,
          votes
        })
      );
    });
};
