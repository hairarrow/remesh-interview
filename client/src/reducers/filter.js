import { RECEIVE_FILTERS } from "../actions/filter";

export default (
  state = { users: [], filters: {}, activeFilters: {} },
  action
) => {
  switch (action.type) {
    case RECEIVE_FILTERS:
      return { ...state, filters: action.filters };
    default:
      return state;
  }
};
