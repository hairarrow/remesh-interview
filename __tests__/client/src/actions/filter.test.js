import configureStore from "redux-mock-store";
import * as filterActions from "../../../../client/src/actions/filter";

const mockStore = configureStore();
const store = mockStore();

describe("filter actions", () => {
  beforeEach(() => {
    store.clearActions();
  });

  test("it receives filters", () => {
    store.dispatch(filterActions.receiveFilters({ age: [] }));
    expect(store.getActions()).toMatchSnapshot();
  });
});
