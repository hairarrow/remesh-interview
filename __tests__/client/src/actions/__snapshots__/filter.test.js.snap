// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter actions it receives filters 1`] = `
Array [
  Object {
    "filters": Object {
      "age": Array [],
    },
    "type": "RECEIVE_FILTERS",
  },
]
`;
